src/OrleansProviders/SQLServer/SqlStatisticsPublisher.cs

```diff
@@ -74,210 +82,122 @@ public void AddConfiguration(string deployment, bool silo, string siloId, SiloAd
+            {
+                logger.Verbose("SqlStatisticsPublisher.ReportMetrics (client) failed: {0}", ex);
+                throw;
             }
         }
 
-        Task ISiloMetricsDataPublisher.Init(string deploymentId, string storageConnectionString, SiloAddress siloAddress, string siloName,
-            IPEndPoint gateway, string hostName)
+
+        async Task ISiloMetricsDataPublisher.Init(string deploymentId, string storageConnectionString, SiloAddress siloAddress, string siloName, IPEndPoint gateway, string hostName)
         {
-            return TaskDone.Done;
+            await InitOrleansQueriesAsync();
         }
 
-        public async Task ReportMetrics(ISiloPerformanceMetrics metricsData)
+
+        public Task ReportMetrics(ISiloPerformanceMetrics metricsData)
         {
-            using (var conn = new SqlConnection(connectionString))
+            if(logger != null) logger.Verbose3("SqlStatisticsPublisher.ReportMetrics (silo) called with data: {0}.", metricsData);
+            try
             {
-                conn.Open();
-                using (var tx = conn.BeginTransaction())
-                {
-                    var command1 = new SqlCommand(METRICS_READ_SINGLE_ROW);
-                    command1.Parameters.Add(new SqlParameter { ParameterName = "@id", DbType = DbType.String, Value = deploymentId });
-                    command1.Parameters.Add(new SqlParameter { ParameterName = "@siloid", DbType = DbType.String, Value = siloName });
-                    command1.Connection = conn;
-                    command1.Transaction = tx;
-
-                    var result = (Int32)await command1.ExecuteScalarAsync();
-
-                    var command2 = new SqlCommand((result > 0) ? METRICS_UPDATE_ROW : METRICS_INSERT_ROW);
-                    ConvertToMetricsRow(metricsData, command2);
-                    command2.Connection = conn;
-                    command2.Transaction = tx;
-
-                    await command2.ExecuteNonQueryAsync();
-                    tx.Commit();
-                }
+                return database.UpsertSiloMetricsAsync(deploymentId, siloName, gateway, siloAddress, hostName, metricsData);
+            }
+            catch(Exception ex)
+            {
+                logger.Verbose("SqlStatisticsPublisher.ReportMetrics (silo) failed: {0}", ex);
+                throw;
             }
         }
 
+
         public async Task ReportStats(List<ICounter> statsCounters)
         {
-            var bldr = new SqlConnectionStringBuilder(connectionString)
-            {
-                MultipleActiveResultSets = true,
-                Pooling = true,
-                AsynchronousProcessing = true
-            };
-
-            using (var conn = new SqlConnection(bldr.ConnectionString))
+            var siloOrClientName = (isSilo) ? siloName : clientId;
+            var id = (isSilo) ? siloAddress.ToLongString() : string.Format("{0}:{1}", siloOrClientName, generation);            
```

Comment:
Maybe these have already a common function somewhere? Maybe formatting these should be refactored to functions of their own and given a descriptive name. It feels like ``id`` isn't a descriptive name for the variable either, but that's how it's all the way to the database currently (in general these variable names in code could be refactored to match exactly the ones in DB, currently they mostly are, but not all).